### DISCUSSION TITLE
Strapi relations: embedding?

### DISCUSSION THREAD

**Discussion by gtamas** [strapi/strapi#6312] [Legacy Topics]
Hi
@lauriejim 

I can see that Strapi supports relations. I'm using mongodb, however, and I would prefer embedding subdocuments and using arrays instead of creating references. Is this possible with Strapi?

So, for example I would like to add comments to my blog post collection this way:

Post:
```
{
   "title":"foobar",
   "subtitle":"something",
   "comments":[
      {
         "title":"comment1",
         "text":"something"
      }
   ]
}
```

How do I create and mange such arrays using the admin panel (or manually)?

As for subdocuments, should I use components to model those?

For example:

Post:
```
{
   "title":"foobar",
   "author":{
       // some object representing the user data. 
   }
}
```

As I understand this, components are also relations (kinda). Data is stored in the db like this:

```
	"tags" : [
		{
			"_id" : ObjectId("5ec6ac40eced33de4f77fa47"),
			"kind" : "ComponentSdssdsdTags",
			"ref" : ObjectId("5ec6ac40eced33de4f77fa40")
		
```

So it's a reference. Is it not possible to create embedded subdocument fields somehow? If the admin can't do it that's fine, but it should be possible to such complex types at least programmatically / manually.

Please help!

EDIT:

If embedding is not possible, could you tell me about about how does Strapi resolve Mongodb references? I mean, do you do multiple selects or is this based on features like the "$lookup" operator?
In other words: how efficient the current implementation is?
Mongo is not a must for us, we can use RDBMS too, although we would prefer Mongo for several reasons. But without the ability of embedding subdocuments this might not best choice.

Thanks in advance!

**Comment by gtamas** [comment]
Anyone?

Please help

**Comment by robertpiosik** [comment]
I would go with SQL and set up an index on relation field in your comments table.

As far as I'm aware Strapi doesn't support embedded documents so it can't utilise MongoDB in a way it was built for and even if it would support them, embedded documents are in most cases a very poor idea which leads to data duplication. You can read more about it in the following article: [Why You Should Never Use MongoDB](http://www.sarahmei.com/blog/2013/11/11/why-you-should-never-use-mongodb/) and in a book [Designing Data-Intensive Applications](https://www.amazon.com/Designing-Data-Intensive-Applications-Reliable-Maintainable/dp/1449373321) which I highly recommend.

**Reply by derrickmehaffy** [reply]
They do not, Strapi effectively treats mongo as a relational database (even though it isn't)

**Comment by gtamas** [comment]
@robertpiosik 
Thanks for the response and the info!

So it seems MongoDB is technically supported by Strapi, but practically it's not, since you can model your data only in a very inefficient way. If this is indeed the case, that's  sad. Maybe the Strapi authors should say Strapi has only limited Mongo support or something like that, because the current situation is very misleading.

As for embedding, I definitely don't agree. It's true that you can't model every kind of data this way, while you can always use an RDBMS, but saying that it's "poor idea in most cases" is kinda an exaggeration.
Actually it works in many cases, and when it does, it provides huge benefits:

Embedding usually results in a simpler data model which will be reflected in your code as well. There will be less code and it's gonna be less complex, so it's gonna be easier to write, change and maintain . Embedded data models might also have huge read performance benefits: instead of performing 7 joins, you can often query your data using a single, simple query which can be many times faster. 
You also don't have to worry about stuff like transactions and (optimistic) locks etc, since you are working with a single entity instead of 7 different tables.

Actually there are many things. Truth is,  MongoDB can be a good choice. Guess that's why it is used by thousands of companies, including tech giants such as Paypal, google, Uber etc. You can find it in 37K+ tach stacks on Strackshare
Obviously all these people use it because it works well for their purposes.

@lauriejim 
Are there any plans for adding better Mongo / document store support? I really think it would important to support embedding as well, since MongoDB works much more efficiently with those models. It would also make easier to add support for other document stores later.

**Reply by robertpiosik** [reply]
I really don't know why mongo is supported whatsoever. As you said, the main idea behind this technology is to properly decide when to embed and eventually duplicate and when to not.

Also, placing mongo among proper databases as interchangeable solutions is misleading.

The first company they list on their front page is facebook which doesn't use mongoDB for years. Unstructured data storage is very helpful in making bigger data sets interpretable but it is considered as a cache then rather a proper solution where data integrity and consistency is a must have. You can read more about MongoDB flaws in [jepsen analysis](http://jepsen.io/analyses/mongodb-4.2.6) - which is a very hot topic in the last days.

I would add the discussion something regarding SQL which is meant to guarantee data consistency by the usage of foreign keys - which are not supported by Strapi.
Fortunately this is an open source project with vibrant community and I believe it will be batter and better as time flies :)

Flaws of database engines usage should be clearly stated in the docs though.

**Reply by derrickmehaffy** [reply]
My personal opinion is that Mongo is not a good choice for Strapi either, it's simply there because those inexperienced in DBA don't understand what mongo's real purpose is (and "buzz words").

Strapi is a strict schema relational framework, neither of which are what mongo was built for. I agree entirely with @robertpiosik 

And yes foreign keys aren't entirely supported yet, though work will start on that with the Strapi v4 (after stable) as far as I'm aware. It's more so just tweaking the generation of the tables, columns in Bookshelf/Knex to assign them.

**Reply by gtamas** [reply]
I still think MongoDB could be a good choice in certain cases, but it requires a different kinda thinking which most people don't seem to like or even understand.
Using RDBMS, you model your data  with 16 tables in *every* case, simply because there is no other way to do it properly. Mongo was designed for the use case, when data could be modelled in the different way. Sometimes it offers an alternative. It doesn't work with all kind of data, but many time it works.
Of course you can use Oracle, MySQL or Postgres for everything, that's also an option.
So in my opinion MongoDB support could be useful for Strapi too, especially since it's a CMS, so it deals with the kind of data MongoDB might be able to handle (depending on how you modelled it  :).

Anyway, the question is not whether Strapi should support Mongo or not. It already does, in some way. But the current level of support is clearly insufficient. I think, proper support should either be implemented to allow MongoDB to be used more efficiently or it should be removed entirely.

**Comment by Marcosul** [comment]
I think it is very important to improve support for using mongodb.

**Comment by lauriejim** [comment]
Hello! because we have to maintain the usage of SQL and NoSQL database, it's quite tricky.
Here is all the code that manages the MongoDB stuff (and relations) - https://github.com/strapi/strapi/tree/master/packages/strapi-connector-mongoose
It's not plan to review the way we are doing relations with MongoDB.

**Comment by alphaadidas** [comment]
I ran into this limitation too, i was hoping 'components' were embeddable.... if they are not embeddable .. then what is the different from a component, vs a join on another entity?  are components just invisible entities?  

Or , would it be possible to have the option for a component to be 'purely embedded' ?   NoSQL supports that, and some SQL db's support object fields too.   Maybe a checkbox that select the behavior, embed or join table.

.... 
@gtamas 
Ultimately, i just added a 'JSON' field to my model/entity, and custom coded some mongoose queries in my service layer.

If you want to maintain some sort of schema for the JSON, you'll need to define custom jsonschema's and validate them each time you update...  or extend the mongooese connector, and figure out how to hack in jsonshcmea's for model json fields. :( 
 ...

Or would it be possible to have another option, when creating a 'json field' ..  add a 'schema to validate with'  property, for the JSON field... ?
